text/0000-min-rust-version.md

```diff
@@ -0,0 +1,198 @@
+- Feature Name: min_rust_version
+- Start Date: 2018-06-28
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add `rust` field to the package section of `Cargo.toml` which will be used to
+specify crate's Minimum Supported Rust Version (MSRV):
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+rust = "1.30"
+```
+
+# Motivation
+[motivation]: #motivation
+
+Currently crates have no way to formally specify MSRV. As a result users can't
+check if crate can be built on their toolchain without building it. It also
+leads to the debate on how to handle crate version change on bumping MSRV,
+conservative approach is to consider such changes as breaking ones, which can
+hinder adoption of new features across ecosystem or result in version number
+inflation, which makes it harder to keep downstream crates up-to-date. More
+relaxed approach on another hand can result in broken crates for user of older
+compiler versions.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`cargo init` will automatically create `Cargo.toml` with `rust` field equal to
+`rust="stable"` or `rust="nightly: *"` depending on the currently used toolcahin.
+On `cargo publish` cargo will take  currently used Rust compiler version and
+will insert it before uploading the crate. In other words localy your `Cargo.toml`
+willl still have `rust="stable"`, but version sent to crates.io will have
+`rust="1.30"` if you've used Rust 1.30. If "nightly: \*" is used, then `cargo`
```

Comment:
I don't like this kind of automatic conversion from `stable` to e.g. `1.30`. I think it is counter-intuitive.

Instead, I would like to propose a slightly more complicated approach with a huge effect on the question raised above (namely "Is a MSRV change a breaking change for a crate which then requires a major version bump?"). Let me give an example:

1. A user might declare the MSRV to be `1.30`. This means that the crate which is published to crates.io can be compiled by every compiler between `1.30` (inclusive) and `2` (exclusive), just as semantic versioning suggests. If he uploads the next (patch or minor) version of the same crate, the MSRV must be equal to or smaller than `1.30`. Everything else would be considered a breaking change. If he uploads the next major version, of course he's free to change the MSRV.

2. A user might also declare the MSRV to be `stable`. When published to crates.io, `stable` is converted to `stable(1.30)` with `1.30` being the current stable release. This means that the published artifact will work with a compiler between version `1.30` (inclusive) and `2` (exclusive). So far no difference! Now, times passes and a new stable compiler (`1.31`) was published. The crate artifact which was already published still works with a compiler between `1.30` (inclusive) and `2` (exclusive). But if the crate author opts to publish a new crate version, the MSRV for the new artifact is automatically raised to `stable(1.31)` which means that it guarantees to work with a compiler version between `1.31` (inclusive) and `2` (exclusive). This MSRV change wound not be considered a breaking change since it was already declared before.

Therefore a crate author has to options:

1. Fix the MSRV to a specific version, e.g. `1.30`.
2. Fix the MSRV to `stable` which means that a future crate version might require a future compiler version (which is at least stable).

Anyway, changing the MSRV is considered to be a breaking change which requires a major version bump. (Someone might relax this requirement as I did in the example above and allow a change from e.g. `stable` to `1.32` or `1.31` to `1.30`, since that is no compatibility problem).

I think, one could take this idea even further and allow labels like `stable - 0.4` which means that, a future crate works with the current version minus 4 minor versions. I think this is even a reasonable alternative to LTS releases as proposed in #2483. (Or we could then say that `lts` is `stable` minus one year or another reasonable time considered to be a long term...)

Reply:
As for LTS releases, they are not only about MSRV, they will also serve as a synchronization point for crate authors and package authors (think Debian). LTS also means that crucial updates (e.g. security or soundness fixes) will be backported to those versions. So I don't think that `stable - 0.4` will be able to cover all LTS use-cases.

Reply:
> Probably my wording was not clear enough, but I've meant almost exactly the same behaviour as was described by you.

To be honest, I am still not sure that we are talking about the same thing: I think, the crucial difference is that in the first case (`rust = 1.30`) the published MSRV on crates.io is `1.30`, while in the second case (`rust = stable`), the published MSRV is `stable(1.30)`. To be even clearer: `1.30` is **not at all** equal to `stable(1.30)`. While they behave equal when it comes to the crate's current version, they are different for the next (patch or minor) version: In the first case the MSRV will be `1.30`, while in the second case it might be `1.30` or `stable(1.31)` or even `1.31`. Therefore, the crate user can determine from the difference between `1.30` and `stable(1.30)` how the crate author answers the question "Is an MSRV change a breaking change?". In the first case, the answer is yes, in the second case it is no!

I would like to see a whole passage dedicated to that matter which leaves no room for interpretation... ;-)

I would also like to see a passage dedicated to "version arithmetic" which covers e.g.:

- Under which circumstances is a MSRV downgrade/upgrade a breaking change?
  - `1.30` can be `1.29`, `1.28.1`, ... in the next crate version.
  - `stable(1.30)` can be `1.29`, `1.28.1`, ... or `stable(1.30.1)`, `stable(1.31)`, ... or `1.30`, `1.30.1`, `1.31`, ... in the next crate version (as long as the rustc is already stabilized). 
- What if crate A depends on crate B and crate C. What are the requirements for the MSRV of crate A given the MSRV of B and C?
  - Something like MSRV(A) <= min(MSRV(A), MSRV(B)). We need a (partial?) order here!
- (How) is it possible to calculate with MSRVs? Examples:
  - stable(1.30.1) - 0.0.1 = stable(1.30)
  - stable(1.30) - 0.3 = stable(1.27)
  - stable(1.30.1) - 0.0.2 = stable(1.30.0) *Does this even work?*
  - stable(1.30) - 1 = undefined

> As for LTS releases, they are not only about MSRV, they will also serve as a synchronization point for crate authors and package authors (think Debian). LTS also means that crucial updates (e.g. security or soundness fixes) will be backported to those versions.

You are right. But I highly doubt the utility of LTS releases for other cases than language version requirements:

- Take the argument about patches that are backported to LTS. If such a package becomes necessary, Debian (or anyone else) will need to adopt the fixed LTS version. In that case, they can also adopt the current stable version, since Rust is guaranteed to be backward-compatible (in version 1.x). LTS releases are even counterproductive in this use case, since they encourage not upgrading to the current release and therefore increase the number of old compilers out there.
- Take the argument about synchronization points. We already have them, they are called "editions". (We could also publish guideline that give suggestion when a good times has come to update and how long crates should support old compiler versions. This could be done in an RFC which depends on this one.)

Reply:
Okay, I've re-read your RFC and I think now that You've mentioned some of my points. But perhaps, they are still useful to clarify things... ;-)

Btw Thanks for Your effort!

Reply:
I was thinking about the theoretical foundations of my proposal and I came up with [this](https://github.com/teiesti/compdep/blob/master/compdep.pdf) article. It is some kind of pre-work for my proposal in this thread but I think its worth sharing here. (Anywhere, I'll try to write another article surrounding these ideas soon.)

An internals thread for a discussion about the article can be found [here](https://internals.rust-lang.org/t/article-compatibility-with-dependencies/7948?u=teiesti).